futures-util/src/task/atomic_waker.rs

```diff
@@ -198,80 +238,91 @@ impl AtomicWaker {
+                let res = self.state.compare_exchange(
+                    REGISTERING, HOLDS, AcqRel, Acquire);
+
+                match res {
+                    Ok(_) => {}
+                    Err(actual) => {
+                        // This branch can only be reached if a
+                        // concurrent thread called `wake`. In this
+                        // case, `actual` **must** be `REGISTERING |
+                        // `WAKING`.
+                        debug_assert_eq!(actual, REGISTERING | WAKING);
+
+                        // Take the waker to wake once the atomic operation has
+                        // completed.
+                        let waker = (*self.waker.get()).take().unwrap();
+
+                        // Just swap, because no one could change state while state == `REGISTERING` | `WAKING`.
+                        self.state.swap(EMPTY, AcqRel);
+
+                        // The atomic swap was complete, now
+                        // wake the task and return.
+                        waker.wake();
                     }
                 }
             }
-            WAKING => {
-                // Currently in the process of waking the task, i.e.,
-                // `wake` is currently being called on the old task handle.
-                // So, we call wake on the new waker
-                lw.wake();
-            }
-            state => {
-                // In this case, a concurrent thread is holding the
-                // "registering" lock. This probably indicates a bug in the
-                // caller's code as racing to call `register` doesn't make much
-                // sense.
-                //
-                // We just want to maintain memory safety. It is ok to drop the
-                // call to `register`.
-                debug_assert!(
-                    state == REGISTERING ||
-                    state == REGISTERING | WAKING);
-            }
+        }
+
+        if state == WAKING || state == WAKING | HOLDS {
+            // Currently in the process of waking the task, i.e.,
+            // `wake` is currently being called on the old task handle.
+            // So, we call wake on the new waker
+            lw.wake();
+        } else {
+            // In this case, a concurrent thread is holding the
+            // "registering" lock. This probably indicates a bug in the
+            // caller's code as racing to call `register` doesn't make much
+            // sense.
+            //
+            // We just want to maintain memory safety. It is ok to drop the
+            // call to `register`.
+            debug_assert!(
+                state == REGISTERING ||
+                state == REGISTERING | WAKING);
         }
     }
 
     /// Calls `wake` on the last `Waker` passed to `register`.
     ///
     /// If `register` has not been called yet, then this does nothing.
     pub fn wake(&self) {
+        let state = self.state.load(Acquire);
```

Comment:
Related to my comment below, but why is it enough to know that the `WAKING` was true, if you also need to know that the thread that is being woken can read your writes up until this function was called?